### PR TITLE
Arm backend: fix round() decomposition to round-half-to-even

### PR DESCRIPTION
**Pull request by vacu9708**
### Summary

`DecomposeRoundPass` lowers `aten.round` with round-half-**away-from-zero**, but `torch.round` is round-half-**to-even** (banker's rounding). (see [`torch.round`](https://pytorch.org/docs/stable/generated/torch.round.html)).

| input | `torch.round` (reference) | delegated output | fixed |
| --- | --- | --- | --- |
| `0.5` | `0` | `1`  | `0`  |
| `2.5` | `2` | `3`  | `2`  |
| `-1.5` | `-2` | `-2`  | `-2`  |
| `-2.5` | `-2` | `-3`  | `-2`  |

### The fix

round(x) now picks whichever of floor(x) and ceil(x) is nearer, and the even one of the two when x is exactly halfway.

```
dist_to_floor = x - floor(x)
halved        = floor(x) * 0.5
floor_is_odd  = (halved - floor(halved)) == 0.5
take_ceil     = (dist_to_floor > 0.5) | ((dist_to_floor == 0.5) & floor_is_odd)
result        = where(take_ceil, ceil(x), floor(x))
```

| `x` | `floor(x)` | `ceil(x)` | `dist_to_floor` | `floor_is_odd` | `take_ceil` | `result` |
| --- | --- | --- | --- | --- | --- | --- |
| `-2.5` | `-3` | `-2` | `0.5` | `true` | `true` | `-2` |
| `-2.25` | `-3` | `-2` | `0.75` | ~~`true`~~ | `true` | `-2` |
| `-1.5` | `-2` | `-1` | `0.5` | `false` | `false` | `-2` |
| `0.49999997` | `0` | `1` | `0.49999997` | ~~`false`~~ | `false` | `0` |
| `0.50000006` | `0` | `1` | `0.50000006` | ~~`false`~~ | `true` | `1` |
| `3.5` | `3` | `4` | `0.5` | `true` | `true` | `4` |

### Testing

| case | data |
| --- | --- |
| `halfway_ties` | exact `.5`, both signs and both parities |
| `halfway_neighbors` | one ulp either side of a tie |

```bash
FILES="backends/arm/test/ops/test_round.py backends/arm/test/misc/test_mixed_type_lowering.py"
PYTEST="pytest --config-file=backends/arm/test/pytest.ini --numprocesses=auto"

$PYTEST $FILES -k tosa            # 17 passed
$PYTEST $FILES -k "u55 or u65"    # 7 passed
$PYTEST $FILES -k u85             # 7 passed

lintrunner backends/arm/_passes/decompose_round_pass.py \
           backends/arm/test/ops/test_round.py \
           backends/arm/test/misc/test_mixed_type_lowering.py
# No lint issues.
```

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218 @mansnils @Sebastian-Larsson @robell @rascani

### CODE REVIEW

**Comment by nil-is-all** [comment]
Thanks for the PR, @vacu9708. Could you address the CI failures and give a ping once it is ready for review?

**Comment by vacu9708** [comment]
Thanks for the review @nil-is-all. This CI failure is caused by `test_mixed_type_lowering_tosa_INT_FP`

`test_mixed_type_lowering_tosa_INT_FP` checks that the right number of Q/DQ nodes are inserted around the excluded op, using hardcoded TOSA op counts per Q and DQ node.

`round()`'s internal decomposition is part of Q's op count, which is changed by this commit.
  
  **Fix:** updated `q_tosa_ops`/`dq_tosa_ops` to the new op counts.

**Comment by nil-is-all** [comment]
> Thanks for the review @nil-is-all. This CI failure is caused by `test_mixed_type_lowering_tosa_INT_FP`
> 
> **What the test does:** `test_mixed_type_lowering_tosa_INT_FP` quantizes a model, producing a mixed FP32/INT32 graph. It checks that:
> 
> * `Sigmoid` stays FP32
> * `Add`/`Mul` stay INT32
> * the right number of Q/DQ nodes are inserted around the excluded op
>   using a hardcoded TOSA op/dtype count per Q and DQ node.
> 
> **Cause:** quantize (`Q`) computes `round(x * inv_scale) + zero_point`, so `round()`'s internal decomposition is part of Q's op count, which is changed by this commit.
> 
> **Fix:** updated `q_tosa_ops`/`dq_tosa_ops` to the new op counts.
> 
> * drops `CEIL`/`GREATER_EQUAL`
> * adds `MUL`/`EQUAL`/`LOGICAL_AND`
> * one extra `FLOOR`/`SUB`

Thanks for the ping, running CI again

**Comment by nil-is-all** [comment]
Hi @vacu9708, could you check the two failing CIs? Once they're sorted, we're good to review & merge.

**Comment by digantdesai** [comment]
@claude review this PR

**Comment by vacu9708** [comment]
Good morning, @nil-is-all
As the Claude bot mentioned, none of the CI failures are related to the changes in this commit or the ARM backend.

**Comment by nil-is-all** [comment]
> Good morning, @nil-is-all As the Claude bot mentioned, none of the CI failures are related to the changes in this commit or the ARM backend.

Thanks for the ping, @vacu9708. @zingo, could you give one final review?

**Comment by zingo** [comment]
Thanks for a good fix! 

Could it be that is_tie also matches values that are merely close to a half-integer? Floating-point rounding might make x + 0.5 integral even when x is not an exact tie. For example:

```
x = torch.nextafter(torch.tensor(0.5), torch.tensor(float("inf")))
torch.round(x)  # 1
x + 0.5        # 1.0
```

It looks like the decomposition might classify this as a tie and return 0. Maybe we could add neighboring-value coverage to check this, something like:

```
def halfway_neighbors():
    ties = torch.tensor([-2.5, -1.5, -0.5, 0.5, 1.5, 2.5])
    return torch.cat(
        (
            torch.nextafter(
                ties, torch.full_like(ties, -float("inf"))
            ),
            torch.nextafter(
                ties, torch.full_like(ties, float("inf"))
            ),
        )
     )

test_data_suite = {
    # Existing cases...
    "halfway_neighbors": halfway_neighbors,
}
```

This might help verify that only exact half-integers use the ties-to-even path.

File: backends/arm/_passes/decompose_round_pass.py [hunk 1]
```diff
@@ -60,26 +37,31 @@ def call_operator(self, op, args, kwargs, meta, updated=False):
...
+        floor = exir_ops.edge.aten.floor.default
+        eq_tensor = exir_ops.edge.aten.eq.Tensor
+        eq_scalar = exir_ops.edge.aten.eq.Scalar
+        logical_and = exir_ops.edge.aten.logical_and.default
+        where = exir_ops.edge.aten.where.self
+
+        x_plus_half = call(add, x, 0.5)
+        rounded_up = call(floor, x_plus_half)
+
+        is_tie = call(eq_tensor, x_plus_half, rounded_up)
```
Comment (zingo) [inline backends/arm/_passes/decompose_round_pass.py]: Could this equality misclassify nearby non-ties after x + 0.5 has already been rounded? I added a general comment with a torch.nextafter reproducer and suggested regression test.

**Review by zingo — CHANGES_REQUESTED** [review]

**Comment by vacu9708** [comment]
Thanks for pointing out the edge case @zingo. Confirmed.

I added a new `halfway_neighbors` test, which checks that values merely close to a half-integer(0.5 ± 1 ulp) round to nearest instead of taking the ties-to-even path.

**Comment by zingo** [comment]
Thanks for the extra fix, I kicked of the testing and will do a proper review when Im at my computer tomorrow. Super thanks for the help!
It will be interesting to see if this will improve something in th Arm Backend flow tests https://github.com/pytorch/executorch/actions/workflows/test-backend-arm.yml

**Comment by zingo** [comment]
Yeah!!! it seem to have fixed the fails with the round() op seen in last nightly ( [test-arm / test-backend-linux (arm_tosa_fp, operators) / linux-job summary](https://github.com/pytorch/executorch/actions/runs/32630062300#summary-97177484471) )

**Comment by vacu9708** [comment]
oh the new test that was recently added in PR #21055 fails with the changes to round().
I'll look into the root cause.

**Comment by zingo** [comment]
I also (in parallel) triggered a re-run  of the test to rule out flakeness

**Comment by zingo** [comment]
> oh the new test that was recently added in PR #21055 fails with the changes to round(). I'll look into the root cause.

I haven't looked into the problem, but It could also be that tests that needs an update.

**Comment by vacu9708** [comment]
Hi @zingo, I've looked into the CI failure and the flow test you mentioned.

## Flow test
I ran the flow test you mentioned locally. This one test changes:
```
backends/test/suite/operators/test_round.py::test_round_values
```
It fails on main and passes with this PR, on both `arm_tosa_fp` and `arm_vgf_fp` tests.

`arm_vgf_fp`, operators suite (467 tests):
| | passed | failed |
|---|---|---|
| main | 378 | 4 |
| this PR | 379 | 3 |

## CI failure
#21055 reduces the number of TRANSPOSE ops by propagating them through the graph until they meet and fuse.

Its current behavior does not support this graph shape `f(x, g(x))`, where a value forks and rejoins:

```mermaid
flowchart LR
  t["TRANSPOSE"] --> x(["x"])
  x --> g["g(x)"]
  x --> f["f(x, g(x))"]
  g --> f
  t -. "cannot move past" .-x f
  style t fill:#fff8c5,stroke:#9a6700,color:#1f2328
  style f fill:#ffd7d5,stroke:#cf222e,color:#1f2328
```

Operators like SiLU (`x * sigmoid(x)`), GELU, etc have this shape, thus they can also be affected by this #21055 issue.

In the failing test (mixed INT+FP), quantization decomposes into arithmetic that includes a round.

- The current `round-half-away-from-zero` is a straight chain without any `f(x, g(x))`, so the test passes.
- The new `round-half-to-even` has `x - floor(x)`, which is `f(x, g(x))`, so it fails.

### Possible fix
If I work on this, the idea would be to move the TRANSPOSE past a whole elementwise region in one step instead of node by node.

```mermaid
flowchart LR
  subgraph R["elementwise region"]
    direction LR
    x(["x"]) --> g["g(x)"]
    x --> f["f(x, g(x))"]
    g --> f
  end
  f --> t["TRANSPOSE"]
  style t fill:#dafbe1,stroke:#1a7f37,color:#1f2328
```

It does not look like a small task, so if this is already on your roadmap I will wait for it. Otherwise I can work on it and send a separate PR.

**Comment by zingo** [comment]
Thanks for the great analysis the TRANSPOSE tests problem is separate and as you mention you just flush it out in the open. I started to see if I can fix that and we might have a fix, lets see. 

You can either wait a few days with this PR and see if we can land a fix before and rebase this ontop or if you want/need it faster just patch that test so it passes and Ill approve that. Then we can enable/fix the tests in the other PR later.

**Comment by zingo** [comment]
Hi my fix might take a few more day , sorry  :( So feel free to not wait on it and just "silent" the failing arm testcase.

**Comment by vacu9708** [comment]
Hi @zingo. I rebased on main and made the two mixed-precision cases XFAIL. They will XPASS once a fix lands and can be dropped in that PR.

**Review by zingo — APPROVED** [review]

**Comment by zingo** [comment]
Approved! But please wait with merging until tests have run to confirm it PASS :)

**Comment by zingo** [comment]
if you dont have permission to merge let me know and I can press the button.

**Comment by vacu9708** [comment]
Could you please merge this? I don't have permission to merge it myself. Thanks!

**Comment by zingo** [comment]
Followup fix for the kind of unrelated test breakage is here https://github.com/pytorch/executorch/pull/22738
